sql/core/src/main/scala/org/apache/spark/sql/execution/joins/RowIterator.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import java.util.NoSuchElementException
+
+import org.apache.spark.sql.catalyst.InternalRow
+
+private[sql] abstract class RowIterator {
+  def advanceNext(): Boolean
+  def getRow: InternalRow
+  def toScala: Iterator[InternalRow] = new RowIteratorToScala(this)
+}
+
+object RowIterator {
+  def fromScala(scalaIter: Iterator[InternalRow]): RowIterator = {
+    scalaIter match {
+      case wrappedRowIter: RowIteratorToScala => wrappedRowIter.rowIter
+      case _ => new RowIteratorFromScala(scalaIter)
+    }
+  }
+}
+
+private final class RowIteratorToScala(val rowIter: RowIterator) extends Iterator[InternalRow] {
+  private [this] var _hasNext: Boolean = rowIter.advanceNext()
+  override def hasNext: Boolean = _hasNext
+  override def next(): InternalRow = {
+    if (!_hasNext) throw new NoSuchElementException
+    val row: InternalRow = rowIter.getRow.copy()
```

Comment:
I'm going to comment this file to explain what's going on here, but in a nutshell:

The Scala iterator exposes separate `hasNext()` and `next()` methods. This creates the potential for a user to call `hasNext()` two times in a row without calling `next()` in between those calls.  For some iterators, like `iterator.filter(...)`, non-trivial work must be performed to figure out whether there's a next tuple and this may cause our mutable row objects to be mutated. This is the reason that we needed the copying in the `.collect()` call in `leftOuterIterator`, even after changing that method to call `collect` on an iterator instead of an iterable.

This wrapper class, `RowIteratorToScala`, makes this safe by adding the required copying at the last possible moment.

In the long run, we can probably remove this copying. The `RowIteratorFromScala` is designed to help with this: that wrapper class takes a Scala iterator and hides it behind our `RowIterator` interface, which prevents the double `hasNext` issue. Therefore, if an operator can use `RowIterators` then it doesn't need to immediately copy its incoming rows. This is the reason that I was able to safely remove the immediate `left.execute.map(_.copy())` and `right.execute.map(_.copy())` calls from SortMergeJoin.